### PR TITLE
Wgan guide

### PR DESCRIPTION
**Pull request by JamesAllingham**
Hey guys! 

Sorry it's taken me so long to make this PR! Life has been getting in the way! 

This is just a first draft and there is still a bunch I need to do, but I figured I should keep you all in the loop with what is going on here. 

Here is what I still need to do:

1. A lot of proofreading, editing, and tweaking. I've done a rough first pass but I still need to read everything thoroughly to find spelling and grammar mistakes and change how things are worded.

2. Add a little more content. Mostly this will be in the form of motivations for some (or all) of the exercises. 

3. Finish making the scribe notes. Currently, there are 7x 30-minute segments that need to be done. I've got volunteers for two of those and I'll find some time to do the rest. Once all of those are done I'll need to combine them, edit them, and link them to this guide. 

4. Incorporate any important questions/answers/points from the notes into the main guide.

5. Incorporate Tali's demos somehow. I still need to talk to him to figure out how best to do this. He made them in Mathematica which most people won't have access too. Some options are to re-implement in another language, use Wolfram CDF, or make them non-interactive. 

6. Tweak the GAN starter notebook to make it a bit easier for people. Maybe add a Pytorch one too (this could maybe come after we publish?). I have added links to Keras implementations of GAN, WGAN, and WGAN-GP for people who get stuck. But I still think it might be nice to have Pytorch because I suspect the implementations for WGAN and especially WGAN-GP will be easier.

And then, of course, I'm sure there will be plenty of feedback from you guys! :) Please let me know if you have any of that, any and all feedback is welcome! You are also welcome to wait until I've done some more proofing before you give feedback. I'll keep updating the PR as I make changes.

### CODE REVIEW

File: _config.yml [hunk 1]
```diff
@@ -27,6 +27,10 @@ krishna:
     name: Kumar Krishna Agrawal
     email: kumarkagrawal@gmail.com
     web: http://twitter.com/kumarkagrawal
+james:
+    name: James Allingham
+    email: 
+    web: https://twitter.com/JamesAllingham
```
Comment (JamesAllingham) [inline _config.yml]: I wasn't sure if I should add my details here but it seemed to work...

File: _posts/2019-03-23-WassersteinGAN.markdown [hunk 2]
```diff
@@ -0,0 +1,522 @@
...
+
+<br />
+
+# 1 Basics of Probability & Information Theory
+  **Motivation**: To understand GAN training (and eventually WGAN & WGAN-GP) we need to first have some understanding of probability and information theory. In particular we will focus on Maximum Likelihood Estimation and the KL-Divergence. This week we will make sure that we understand the basics so that we can build upon them in the following weeks.  
+
+  _This week contains some fairly introductory material, if everyone participating in this curriculum is comfortable with this material you may wish to treat this week as optional or as a pre-requisite. However, this week still covers some very interesting and cool topics, and it is important to have a solid grasp of these concepts in order to build towards understanding the Wasserstein GAN, so skip with caution._
+
```
Comment (JamesAllingham) [inline _posts/2019-03-23-WassersteinGAN.markdown]: Not super sure about this comment. It seemed like many people in the group found the first week a bit dull (too introductory) so I wanted to have some kind of warning that people might want to give it a skip. However, I think that a lot of the concepts from this week are not only very interesting but also important for the following weeks so I don't want to encourage too many people to skip it!

File: _posts/2019-03-23-WassersteinGAN.markdown [hunk 3]
```diff
@@ -0,0 +1,522 @@
...
+  **Notes**: Here is a [link](todo:add_link) to our notes for the lesson. We were fortunate enough to have Tim Salimans sit in on the session!
+
+  <br />
+
+# 3 Generative Adversarial Networks
+
+  **Motivation**: Let’s read the original GAN paper. Our main goal this week is to understand how GANs solve some of the problems with training generative models, as well as, some of the new issues that come with training GANs.
+
+  _This week has a fairly heavy workload and could plausibly be split over two weeks. This is due to having two papers for required reading. However, it should be noted that the second paper is really somewhere between required and optional &#8212; we think that it contains some  interesting material and sets the state well for looking at WGAN in week 4, but the most important points are covered again in week 4. Depending on your study group you might want to focus on this paper more or less based on your interest (we recommend that most groups don't spend too much time on this paper)._
+
```
Comment (JamesAllingham) [inline _posts/2019-03-23-WassersteinGAN.markdown]: Again, this comment was a bit tricky. Don't want to discourage people from reading the second paper, it is very interesting and everyone from the group who read it got pretty engrossed by it. However, as a result of people getting a little sidetracked the week became a bit overloaded so I wanted to give some kind of warning.

**Comment by suryabhupa** [comment]
Hey James! I'll spend some time going over the curriculum on the website in a bit more detail and provide feedback as it comes! 

1. Sounds good, how do you want us to help you with this? I can request changes to the PR if that'd be helpful.

2. I think the motivations are an important part of the section so making sure it's well-written is important. 

3. 👍 Thanks for doing this! 

4. 👍 

5. Yeah, that's a bit tricky. My first instinct would be to have them be in Colab, but it's up to you to decide what you mean by interactivity -- does a code sample with fill-in-the-blanks fulfill this, or would you rather have cool interactive js diagrams? Up to you (though Colabs are always a safe bet). 

6. Feel free to add the PyTorch one later, because I think it's probably good as it stands. We can first prioritize getting the Mathematica --> Colabs, and the scribe notes, and then adding a PyTorch implementation. 

I'll start editing the PR over the next few days.

File: _posts/2019-03-23-WassersteinGAN.markdown (hunk 2 above)
Comment (suryabhupa) [reply _posts/2019-03-23-WassersteinGAN.markdown]: You can feel free to keep this, but also keep in mind that more often than not, I imagine a large part of the target audience will be individuals doing this on a self-paced time. 

I'd totally support you removing this comment entirely if you wanted, but also I think it's fine if you want to keep it. What do you think?

File: _posts/2019-03-23-WassersteinGAN.markdown (hunk 3 above)
Comment (suryabhupa) [reply _posts/2019-03-23-WassersteinGAN.markdown]: As opposed to the first comment, which is just highlighting that the first week has fairly basic (yet important fundamental) stuff, I think this comment is pretty fair. You might be able to make this comment less verbose and just highlight the second paper is optional, but highly recommended? Curious to hear your thoughts.

File: _config.yml (hunk 1 above)
Comment (suryabhupa) [reply _config.yml]: Cool, go for it, haha.

**Comment by JamesAllingham** [comment]
1. That sounds great. I'm going to find some time to do some editing in the next 24 hours, so after that, any additional changes you can recommend would be very helpful - before then you'll probably just be finding many small and silly mistakes.

2. I agree. I'll also find some time to do the first draft of those in the next 24 hours.

5. Cool, will get back to you about this ASAP :) 

6. Okay, that seems like a reasonable way to prioritize this.

Thanks!

File: _posts/2019-03-23-WassersteinGAN.markdown (hunk 3 above)
Comment (JamesAllingham) [reply _posts/2019-03-23-WassersteinGAN.markdown]: I agree that it was a bit too verbose. Let me know what you think of the changes I made.

File: _posts/2019-03-23-WassersteinGAN.markdown (hunk 2 above)
Comment (JamesAllingham) [reply _posts/2019-03-23-WassersteinGAN.markdown]: I'm still undecided about this. I am leaning towards just removing it. For now, I've left it in though. I think that if most people are doing this curriculum in a self-paced setting, they can pretty quickly figure out which material is too introductory for them and then just skip it - we don't need to tell them :)

File: _posts/2019-03-23-WassersteinGAN.markdown [hunk 4]
```diff
@@ -0,0 +1,528 @@
...
+  2. Ch 3.13 of the DL book
+      * This chapter covers KL-Divergence & the idea of distances between probability distributions which will also be a key concept going forward.
+  3. Chs 5.1.4 and 5.5 of the DL book
+      * The aim of these chapters is to make sure that everyone understands Maximum Likelihood Estimation which is a fundamental concept in machine learning. It is used explicitly or implicitly in both supervised and unsupervised learning as well as in both discriminative and generative methods. In fact, many methods using gradient descent are doing approximate MLE. It is important to understanding MLE as a fundamental concept, and its use in machine learning in practice. Note that, if you are not familiar with the notation used in these chapters, you might want to start at the beginning of the chapter. Also note that, if you are not familiar with the concept of estimators, you might want to read Ch 5.4. However, you can probably get by simply knowing that minimizing MSE is a method for optimizing some approximation for a function we are trying to learn (an estimator).
```
Comment (suryabhupa) [inline _posts/2019-03-23-WassersteinGAN.markdown]:
```suggestion
      * The aim of these chapters is to make sure that everyone understands maximum likelihood estimation (MLE) which is a fundamental concept in machine learning. It is used explicitly or implicitly in both supervised and unsupervised learning as well as in both discriminative and generative methods. In fact, many methods using gradient descent are doing approximate MLE. It is important to understanding MLE as a fundamental concept, and its use in machine learning in practice. Note that, if you are not familiar with the notation used in these chapters, you might want to start at the beginning of the chapter. Also note that, if you are not familiar with the concept of estimators, you might want to read Ch 5.4. However, you can probably get by simply knowing that minimizing mean squared error (MSE) is a method for optimizing some approximation for a function we are trying to learn (an estimator).
```

File: _posts/2019-03-23-WassersteinGAN.markdown [hunk 5]
```diff
@@ -0,0 +1,528 @@
...
+     * An alternative discussion on the links between MLE and MSE.
+  5. The first 37ish minutes of Arthur Gretton's MLSS Africa talk on comparing probability distributions &#8212; [video](https://www.youtube.com/watch?v=5sijxSg8P14), [slides](https://drive.google.com/file/d/1RNrgDs5xw-9HTjikFU1L0iO1PBMDaGwE/view)
+     * An interesting take on comparing probability distributions. The first 37 minutes are fairly general and give some nice insights as well as some foreshadowing of what we will be covering in the following weeks. The rest of the talk is also very interesting and ends up covering another GAN called the MMD-GAN, but it isn’t all that relevant for us.
+  6. [On Integral Probability Metrics, φ-Divergences and Binary Classification](https://pdfs.semanticscholar.org/6af2/fa8887a2cb0386f79e3a2822b661e2dc8369.pdf)
+     * For those of you whose curiosity was peaked by Arthur’s talk, this paper goes into depth describing IPMs (such as MMD and the 1-Wasserstein distance) and comparing them the φ-divergences (such as the KL-Divergence). *This paper is fairly heavy mathematically so don't be discouraged if you struggle to follow it*.
```
Comment (suryabhupa) [inline _posts/2019-03-23-WassersteinGAN.markdown]:
```suggestion
     * For those of you whose curiosity was piqued by Arthur’s talk, this paper goes into depth describing IPMs (such as MMD and the 1-Wasserstein distance) and comparing them the φ-divergences (such as the KL-Divergence). *This paper is fairly heavy mathematically so don't be discouraged if you struggle to follow it*.
```

File: _posts/2019-03-23-WassersteinGAN.markdown [hunk 6]
```diff
@@ -0,0 +1,528 @@
...
+     If we divide this term by a constant factor of \(N\) we the same term that would minimize the to maximize the KLD: \(\mathbb{E}_p[\log q(x|\bar{\theta})]\).
+     </p>
+     </details>
+
+**Notes**: Here is a [link](todo:add_link) to our notes for the lesson. We were fortunate enough to have Martin Arjovsky sit in on the session!
+
+<br />
+
+# 2 Generative Models
+  **Motivation**: This week we’ll take a look at generative models. We will aim to understand how they are similar and how they differ from discriminative models. In particular, we want to understand the challenges that come with training generative models.
```
Comment (suryabhupa) [inline _posts/2019-03-23-WassersteinGAN.markdown]:
```suggestion
  **Motivation**: This week we’ll take a look at generative models. We will aim to understand how they are similar and how they differ from the discriminative models covered last week. In particular, we want to understand the challenges that come with training generative models.
```

File: _posts/2019-03-23-WassersteinGAN.markdown [hunk 7]
```diff
@@ -0,0 +1,528 @@
...
+      </p>
+      </details>
+
+  2. Describe in your own words the difference between a generative and a discriminative model.
+      <details><summary>Solution</summary>
+      <p>
+      This is an open ended question but here are some of the differences:
+      <ul>
+        <li>In the generative setting, we usually model \(p(x)\), our models are usually non-deterministic, and we can sample from them.</li>
+        <li>In the discriminative setting, we usually model \(p(y|x)\), our models are often deterministic, and we can't necessarily sample from them.</li>
```
Comment (suryabhupa) [inline _posts/2019-03-23-WassersteinGAN.markdown]: I would also add something to the effect of discriminative models typically need labels, generative models don't need them, as their objective is different.

File: _posts/2019-03-23-WassersteinGAN.markdown [hunk 8]
```diff
@@ -0,0 +1,528 @@
...
+
+  **Topics**:
+
+  1. Wasserstein Distance vs KLD/JSD
+  2. Wasserstein GAN
+
+  **Required Reading**:
+
+  1. [The WGAN paper](https://arxiv.org/pdf/1701.07875.pdf)
+      * This should be pretty self-explanatory! We’re doing a DFL on Wasserstein GANs so we’d better read the paper! (This isn’t the end of the road, however, next week we’ll look at WGAN-GP.) The paper builds upon an intuitive idea: the family of Wasserstein distances is a nice distance between probability distributions, that is well grounded in theory. The authors propose to use the 1-Wasserstein distance to estimate generative models. They show that the 1-Wasserstein distance is an IPM with a meaningful set of constraints (1-Lipschitz functions), and can, therefore, be optimized by focusing on discriminators that are “well behaved” (meaning that their output does not change to much if you perturb the input, i.e. they are Lipschitz!).
```
Comment (suryabhupa) [inline _posts/2019-03-23-WassersteinGAN.markdown]: I think the statement: "use the 1-Wasserstein distance to estimate generative models" should be expanded on a little bit.

Comment (suryabhupa) [inline _posts/2019-03-23-WassersteinGAN.markdown]:
```suggestion
      * This should be pretty self-explanatory! We’re doing a DFL on Wasserstein GANs so we’d better read the paper! (This isn’t the end of the road, however, next week we’ll look at WGAN-GP.) The paper builds upon an intuitive idea: the family of Wasserstein distances is a nice distance between probability distributions, that is well grounded in theory. The authors propose to use the 1-Wasserstein distance to estimate generative models. They show that the 1-Wasserstein distance is an integral probability metric (IPM) with a meaningful set of constraints (1-Lipschitz functions), and can, therefore, be optimized by focusing on discriminators that are “well behaved” (meaning that their output does not change to much if you perturb the input, i.e. they are Lipschitz!).
```

Comment (JamesAllingham) [reply _posts/2019-03-23-WassersteinGAN.markdown]: I've expanded a bit on this in my most recent commits, let me know if you think it needs more.

File: _posts/2019-03-23-WassersteinGAN.markdown (hunk 7 above)
Comment (JamesAllingham) [reply _posts/2019-03-23-WassersteinGAN.markdown]: Thanks, I added this.

File: _posts/2019-03-23-WassersteinGAN.markdown (hunk 2 above)
Comment (JamesAllingham) [reply _posts/2019-03-23-WassersteinGAN.markdown]: I decided to remove this comment.

**Comment by JamesAllingham** [comment]
Okay, I've incorporated most of (all?) of your feedback and I've also added the notes from our sessions! I think we can take this live now! 🎈 

There are some things left to do:

1. Add Pytorch solution notebooks. As discussed before, we can launch without these and add them later since we do already have Keras solutions.
2. Add Tali's stuff. Having looked at the recordings and spoken to him about this, I think we probably overvalued this. It won't end up being an interactive thing, but rather some cool, but static, visualizations in the notes for that week. I don't think it is worth delaying for this.
3. Go through the notes and generalize them in some places so that they are more useful for folks who weren't in the sessions with us. I think the notes are still useful for the most part, just that they could be slightly improved in some cases, so let's not wait on this either IMO. 🎉 

Let me know what you think, but I am happy and excited to launch this! 

Oh, one other thing that needs to happen before launching is to change the date in the name of the markdown file as well as in the header at the top 😸

**Comment by suryabhupa** [comment]
This all looks quite good! I'm quite pleased with it now -- fantastic work, James! :) I'll go ahead and do the merge now.

**Comment by suryabhupa** [comment]
@JamesAllingham Can you just do final checks and make sure it renders properly? Let's close this after that.

**Comment by suryabhupa** [comment]
+ as well as finalize all acknowledgements.
